R/peaks-functions.R

```diff
@@ -429,3 +429,65 @@ joinPeaksGnps <- function(x, y, xPrecursorMz = NA_real_,
     x[, 2L] <- MsCoreUtils::smooth(x[, 2L], cf = coef)
     x
 }
+
+#' @description
+#'
+#' Removes (Orbitrap) FFT artefacts (peaks) from a spectrum keeping peaks
+#' potentially representing [13]C isotopes.
+#'
+#' @param x peaks matrix
+#'
+#' @param halfWindowSize `numeric(1)` defining the (m/z) window left and right
+#'     of a peak to look for artefacts.
+#'
+#' @param threshold `numeric(1)` defining the relative intensity to the actual
+#'     peak below which peaks are considered artefacts. Defaults to
+#'     `threshold = 0.2` hence removing all peaks with an intensity below 0.2 *
+#'     the peak intensity.
+#'
+#' @param keepIsotopes `logical(1)` whether it should be screened for potential
+#'     isotope peaks. Defaults to `keepIsotopes = TRUE` thus candidate artefact
+#'     peaks potentially being isotopes will **not** be removed.
+#'
+#' @param maxCharge `integer(1)`
+#'
+#' @param isotopeTolerance `numeric(1)`
+#'
+#' @author Jan Stanstrup
+#'
+#' @importFrom MsCoreUtils between
+#'
+#' @noRd
+.peaks_remove_fft_artifact <- function(x, halfWindowSize = 0.05,
+                                       threshold = 0.2,
+                                       keepIsotopes = TRUE,
+                                       maxCharge = 5,
+                                       isotopeTolerance = 0.005) {
+    neutron   <- 1.0033548378 # really C12, C13 difference
+    iso_dist  <- neutron / seq(from = 1, by = 1, to = maxCharge)
+    find_isotopes <- keepIsotopes & any(iso_dist < halfWindowSize)
+    mz <- x[, "mz"]
+    int <- x[, "intensity"]
+    idx <- order(int, decreasing = TRUE)
+    keep <- rep(TRUE, length(idx))
+    while (!is.na(i <- idx[1L])) {
+        idx <- idx[-1L]
+        rem_candidate <- which(
+            between(mz[idx], mz[i] + c(-halfWindowSize, halfWindowSize)))
+        rem_candidate <- rem_candidate[int[idx][rem_candidate] / int[i] <
+                                       threshold]
+        ## That part might be improved later.
+        if (find_isotopes) {
+            target_dist <- abs(mz[idx][rem_candidate] - mz[i])
+            dist_matrix <- outer(target_dist, iso_dist, "-")
+            dist_matrix <- abs(dist_matrix)
+            dist_to_iso_hypo <- apply(dist_matrix, 1, min)
+            rem_candidate <- rem_candidate[dist_to_iso_hypo > isotopeTolerance]
+        }
+        if (length(rem_candidate)) {
+            keep[idx[rem_candidate]] <- FALSE
+            idx <- idx[-rem_candidate]
+        }
+    }
```

Comment:
I like to suggest an alternative implementation here.
1. I would replace the `find_isotopes` part with `MsCoreUtils::common` (internally `MsCoreUtils::closests`).
2. I would take the index search outside the loop. In this context I would replace `MsCoreUtils::between` by two `findInterval` calls.

```suggestion
    ## just calculate isotopes that are in the halfWindowSize
    iso_dist <- iso_dist[iso_dist < halfWindowSize]
    find_isotopes <- keepIsotopes & length(iso_dist)
    if (find_isotopes)
        iso_dist <- c(-iso_dist, rev(iso_dist))

    mz <- x[, "mz"]
    int <- x[, "intensity"]

    ## left boundary
    lb <- findInterval(mz - halfWindowSize, mz) + 1L
    ## right boundary
    rb <- findInterval(mz + halfWindowSize, mz)

    ## region of interest (we just need to test if the window spans more than 1
    ## index)
    roi <- rb > lb

    ## test from the largest intensity
    idx <- seq_along(int)[roi][order(int[roi], decreasing = TRUE)]
    keep <- rep(TRUE, length(mz))

    for (i in idx) {
        if (keep[i]) {
            rem_candidate <- seq.int(lb[i], rb[i], by = 1L)
            rem_candidate <-
                rem_candidate[int[rem_candidate] / int[i] < threshold]

            if (find_isotopes) {
                cmm <- common(
                    mz[rem_candidate],
                    mz[i] + iso_dist,
                    tolerance = isotopeTolerance,
                    duplicates = "closest",
                    .check = FALSE
                )
                rem_candidate <- rem_candidate[!cmm]
            }

            keep[rem_candidate] <- FALSE
        }
    }
```

This changes yield a 5~6 runtime improvement in the aforementioned examples:
```r
library("Spectra")
library("MsCoreUtils")

.peaks_remove_fft_artifact2 <- function(x, halfWindowSize = 0.05,
                                        threshold = 0.2,
                                        keepIsotopes = TRUE,
                                        maxCharge = 5,
                                        isotopeTolerance = 0.005) {
    neutron   <- 1.0033548378 # really C12, C13 difference

    iso_dist <- neutron / seq.int(from = 1L, to = maxCharge, by = 1L)

    ## just calculate isotopes that are in the halfWindowSize
    iso_dist <- iso_dist[iso_dist < halfWindowSize]
    find_isotopes <- keepIsotopes & length(iso_dist)
    if (find_isotopes)
        iso_dist <- c(-iso_dist, rev(iso_dist))

    mz <- x[, "mz"]
    int <- x[, "intensity"]

    ## left boundary
    lb <- findInterval(mz - halfWindowSize, mz) + 1L
    ## right boundary
    rb <- findInterval(mz + halfWindowSize, mz)

    ## region of interest (we just need to test if the window spans more than 1
    ## index)
    roi <- rb > lb

    ## test from the largest intensity
    idx <- seq_along(int)[roi][order(int[roi], decreasing = TRUE)]
    keep <- rep(TRUE, length(mz))

    for (i in idx) {
        if (keep[i]) {
            rem_candidate <- seq.int(lb[i], rb[i], by = 1L)
            rem_candidate <-
                rem_candidate[int[rem_candidate] / int[i] < threshold]

            if (find_isotopes) {
                cmm <- common(
                    mz[rem_candidate],
                    mz[i] + iso_dist,
                    tolerance = isotopeTolerance,
                    duplicates = "closest",
                    .check = FALSE
                )
                rem_candidate <- rem_candidate[!cmm]
            }

            keep[rem_candidate] <- FALSE
        }
    }
    x[keep, , drop = FALSE]
}

data(fft_spectrum)
library(microbenchmark)
sp <- peaksData(fft_spectrum)[[1L]]
microbenchmark(
    Spectra:::.peaks_remove_fft_artifact(sp),
    .peaks_remove_fft_artifact2(sp),
    check = "identical"
)
# Unit: microseconds
#                                      expr      min        lq      mean
#  Spectra:::.peaks_remove_fft_artifact(sp) 2934.262 2972.7795 3527.2572
#           .peaks_remove_fft_artifact2(sp)  421.241  446.2135  510.7289
#   median        uq      max neval
#  3022.68 3170.3490 9368.680   100
#   527.62  561.3365  657.874   100

## with isotope checking
microbenchmark(
    Spectra:::.peaks_remove_fft_artifact(sp, halfWindowSize = 1),
    .peaks_remove_fft_artifact2(sp, halfWindowSize = 1),
    check = "identical"
)

# Unit: milliseconds
#                                                          expr      min
#  Spectra:::.peaks_remove_fft_artifact(sp, halfWindowSize = 1) 8.331218
#           .peaks_remove_fft_artifact2(sp, halfWindowSize = 1) 2.087970
#        lq     mean   median       uq       max neval
#  8.535107 9.114861 8.710721 8.873770 13.359157   100
#  2.226285 2.386233 2.294129 2.394645  7.202417   100

sciex_file <- normalizePath(
    dir(system.file("sciex", package = "msdata"), full.names = TRUE))
sciex_mzr <- backendInitialize(MsBackendMzR(), files = sciex_file)
sp <- peaksData(sciex_mzr)[[1705L]]
microbenchmark(
    Spectra:::.peaks_remove_fft_artifact(sp),
    .peaks_remove_fft_artifact2(sp),
    check = "identical"
)
# Unit: microseconds
#                                      expr      min        lq      mean
#  Spectra:::.peaks_remove_fft_artifact(sp) 2798.627 2886.6220 3355.8434
#           .peaks_remove_fft_artifact2(sp)  401.261  418.6355  487.2826
#     median       uq      max neval
#  2970.8425 3067.638 8554.893   100
#   467.0105  517.454 1283.004   100
```

Reply:
And again, @sgibb comes out of nowhere and between two nigh shifts proposes an alternative implementation that substantially speeds things up. So impressive!

Reply:
That is indeed damn impressive and a really meaningful improvement!

Reply:
Great indeed @sgibb ! Only, I would change `duplicates = "closest"` to `duplicates = "keep"` - we've still some issues with the *closest* option (there is an issue is `MsCoreUtils`), and we also want to be a little bit more permissive keeping all possible isotope peaks.